fix(file tree): ls-tree if no grep

gerhardol opened the pull request:
fixes #12413 
fixes #12464 

## Proposed changes

Improve performance (especially for lfs files) to display the file tree if no grep filter is active.
See the issue for the reduced time to run the command.

This could be moved to GitModule instead.
GitModule is a better place for checking that 

Note: ls-tree cannot be used for artificial commits, but ls-files could be used
That is a separate optimization

## Test methodology <!-- How did you ensure quality? -->

view manual git log

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented on src/app/GitUI/UserControls/FileStatusDiffCalculator.cs:
```diff
@@ -315,7 +315,9 @@ GitRevision GetActualRevisionForRevision(GitRevision revision)
 
             IGitModule module = GetModule();
-            IReadOnlyList<GitItemStatus> statuses = module.GetGrepFilesStatus(selectedRev.ObjectId, _fileStatusDiffCalculatorInfo.GrepArguments, applyAppSettings: !_fileStatusDiffCalculatorInfo.FileTreeMode, cancellationToken);
+            IReadOnlyList<GitItemStatus> statuses = !selectedRev.IsArtificial && _fileStatusDiffCalculatorInfo.GrepArguments == @"-e ""^"""
+                ? module.GetTreeFiles(selectedRev.ObjectId, full: true)
+                : module.GetGrepFilesStatus(selectedRev.ObjectId, _fileStatusDiffCalculatorInfo.GrepArguments, applyAppSettings: !_fileStatusDiffCalculatorInfo.FileTreeMode, cancellationToken);
```
Will cleanup a little: Remove the hardcoded setting of GrepArguments  to ^ for file tree (use "") and move that to GetTreeFiles() for articicial (replace later with ls-files?) (maybe cancellation token too, but ls-tree does not take more than 200ms also in Linux repo so not long running).

mstv replied:
Thank you, and much simpler than feared. 🎉

mstv reviewed (commented):
The major change LGTM.
It needs two additions:
29d98228c48eb38b191caf4cadb0f0905c037501
dc2d21cf58a2ada9443631c6da10d664e6c1230d

gerhardol commented:
There is something still that is not right,  but it probably works:
If grep status is set once, icon is updated from that on without manipulating the status (why I got it working without the grep string). But GitItemStatus is set with  prefix:
                            summary: $"{_grepSummaryPrefix}{_fileStatusDiffCalculatorInfo.GrepArguments} {GetDescriptionForRevision(selectedRev.ObjectId)}",

mstv commented:
> There is something still that is not right, but it probably works: If grep status is set once, icon is updated from that on without manipulating the status (why I got it working without the grep string).

This is not wrong: Icons are cached. If they have already been loaded, they can be used at once.
The loading of icons is triggered for git-grep results only because actual *status* items have their special icon.

> But GitItemStatus is set with prefix: summary: $"{_grepSummaryPrefix}{_fileStatusDiffCalculatorInfo.GrepArguments} {GetDescriptionForRevision(selectedRev.ObjectId)}",

How is this related to this PR?
This `summary` is used for diff groups (`FileStatusWithDescription`) but not for single items (`GitItemStatus`).
For file tree, there is only one "group" without heading.

mstv commented on src/app/GitUI/UserControls/FileStatusDiffCalculator.cs:
```diff
@@ -83,21 +83,7 @@ private List<FileStatusWithDescription> CalculateDiffs(
-                            : module.GetTreeFiles(selectedRev.TreeGuid, full: true)));
-                }
-                else
+                if (actualRev.ParentIds?.Count is > 0)
```
This is just a refactoring (condition equivalently inverted [and shortened]). Or have I missed something?

gerhardol replied:
correct, added when looking at it - I can revert

gerhardol commented:
submodules are not displayed correctly after my latest change

gerhardol commented on src/app/GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1436,9 +1436,12 @@ int GetItemImageIndex(GitItemStatus gitItemStatus, bool isGitGrep)
-                    ? gitItemStatus == noItemStatuses[0] && !isGitGrep ? nameof(Images.FileStatusCopiedSame) : nameof(Images.FileStatusUnknown)
-                    : GetItemImageKey(gitItemStatus);
+                // FileTree/grep images are added async to _imageListData
+                string imageKey = isGitGrep
+                    ? nameof(ImageListData.DefaultFileImage)
```
The change is just to set the placeholder icon explicitly, without calling GetItemImageKey() and rely on the fake GitItemStatus info.
Cached image is handled a few lines before,.

For submodules GetItemImageKey() will need to be called though, so a little more to handle (in addition to inserting submodules) why I set this as a draft. (maybe not handled here)
Maybe go back to setting fake grep string.

Maybe adding ls-files for artificial commits too, that seem to be quite minor and some more  special handling can be removed.

gerhardol commented:
Two commits added, left a branch behind and squashed and rebased
ls-files works fine, could remove the special submodule handling too. Some optimizations, probably not possible to see. (I belive that there are no additional async added.)

mstv commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -3191,18 +3185,35 @@ public IReadOnlyList<string> GetFullTree(string id)
-                { full, "-r" },
-                commitId
-            };
+                throw new ArgumentOutOfRangeException(nameof(full), "Artificial commit requires fu.");
```
There is no need to leave a nit for me. :)
```suggestion
                throw new ArgumentOutOfRangeException(nameof(full), "Artificial commit requires 'full'.");
```

mstv commented on src/app/GitCommands/Git/GitTreeParser.cs:
```diff
@@ -33,10 +33,13 @@ public IEnumerable<GitItem> Parse(string? tree)
+            IEnumerable<string> items = tree.LazySplit('\0');
 
-            return items.Select(ParseSingle).Where(item => item is not null)!;
+            return from item in items
+                   let parsed = ParseSingle(item)
+                   where parsed is not null
+                   select parsed;
```
Do we (really) want to switch to this syntax? I would prefer at least that C# would add an implicit `select` clause because it is pretty redundant. What else should be meant?

gerhardol replied:
I asked CoPilot that suggested that this more efficient, less allocations as it iterates.
A foreach should be the same.
Why the LINQ is not iterating by default is something I do not realize. 
@russkie?

RussKie replied:
The framework teams don't use the linq at all (in the runtime implementations), as its perf isn't there (it's being improved, but there are still better alternatives).
I personally never like the sql-style linq syntax, and prefer the original syntax (or `foreach`, if applicable).

RussKie replied:
> I asked CoPilot that suggested that this more efficient, less allocations as it iterates.

It can and does hallucinate every now and then :)

mstv reviewed (commented):
works well and fast for me, have not tested with removed submodule
In `GetLsTreeOutput` for `GitTreeParserTests.Parse_should_return_the_list`, the `'\n'` need to be replaced with `'\0'`.

gerhardol commented:
> have not tested with removed submodule

A removed subm is not included in the output, except for git ls-files --no-cached
The attempt to get data is obviously failing
This is a Git issue, ignore

![image](https://github.com/user-attachments/assets/d87da36c-cec1-4f1d-9b41-1612d88ce959)

Before the subm information could be incorrect, the following lists also the removed 

HEAD
![image](https://github.com/user-attachments/assets/a11ab8ea-e322-4b3d-b05e-1e4bcee64299)

index
it is a bug to attempt getting the treeid here
![image](https://github.com/user-attachments/assets/308393d3-ced2-4c96-8f61-6bef919e606a)

worktree
![image](https://github.com/user-attachments/assets/3e0c292c-1d1f-4c22-8207-254da58a90e9)

--
In diff the behavior is the same as before
New submodule is wrong, but the subm info is not available

![image](https://github.com/user-attachments/assets/8c327f72-5579-4604-a477-516f1e1375cd)

mstv reviewed (approved):
If it is actually more efficient than just
`return items.Select(ParseSingle).WhereNotNull();`

(The implementation of `WhereNotNull` could be turned into `Where(item => item is not null)` in order to let the IL compiler optimize it. I guess there is also a cost to the statemachine for `IEnumerable` and `yield`.)

gerhardol commented:
squashed, rebased and added https://github.com/gitextensions/gitextensions/pull/12436#discussion_r2260879672
Plan to merge tomorrow together with #12421
